chunked_stream/lib/src/chunked_stream_iterator.dart

```diff
@@ -23,197 +24,207 @@ import 'dart:math';
-  /// error occurred.
+  /// If [read] is called before the sub-[Stream] is fully read, a [StateError]
+  /// will be thrown.
   ///
-  /// This method may not be called concurrently.
-  Future<List<T>> read(int size) async {
-    ArgumentError.checkNotNull(size, 'size');
-    if (size <= 0) {
-      throw ArgumentError.value(size, 'size', 'size must be greater than zero');
-    }
-    if (_wakeup != null) {
-      throw StateError('Concurrent invocations not supported');
-    }
+  /// ```dart
+  /// final s = ChunkedStreamIterator(_chunkedStream([
+  ///   ['a', 'b', 'c'],
+  ///   ['1', '2'],
+  /// ]));
+  /// expect(await s.read(1), equals(['a']));
+  ///
+  /// // creates a substream from the chunks holding the
+  /// // next three elements (['b', 'c'], ['1'])
+  /// final i = StreamIterator(s.substream(3));
+  /// expect(await i.moveNext(), isTrue);
+  /// expect(await i.current, equals(['b', 'c']));
+  /// expect(await i.moveNext(), isTrue);
+  /// expect(await i.current, equals(['1']));
+  ///
+  /// // Since the substream has been read till the end, we can continue reading
+  /// // from the initial stream.
+  /// expect(await s.read(1), equals(['2']));
+  /// ```
+  ///
+  /// The resulting stream may contain less than [size] elements if the
+  /// underlying stream has less than [size] elements before the end of stream.
+  ///
+  /// When the substream is cancelled, the remaining elements in the substream
+  /// are drained.
+  Stream<List<T>> substream(int size);
+}
 
-    final c = Completer<List<T>>();
-    _wakeup = () {
-      if (_lastChunk != null) {
-        _buffered.addAll(_lastChunk);
-        _lastChunk = null;
-      }
-      if (_buffered.length >= size) {
-        c.complete(_buffered.sublist(0, size));
-        _buffered = _buffered.sublist(size);
-        _reset();
-        return;
-      }
-      if (_error != null) {
-        c.completeError(_error, _stackTrace);
-        _error = null;
-        _stackTrace = null;
-        _reset();
-        return;
-      }
-      if (_isDone) {
-        c.complete(_buffered);
-        _buffered = [];
-        _reset();
-        return;
-      }
-    };
-    _wakeup();
-    if (!c.isCompleted) {
-      _subscription.resume();
+/// General purpose _chunked stream iterator_.
+class _ChunkedStreamIterator<T> implements ChunkedStreamIterator<T> {
+  /// Underlying iterator that iterates through the original stream.
+  final StreamIterator<List<T>> _iterator;
+
+  /// Keeps track of the number of elements left in the current substream.
+  int _toRead = 0;
+
+  /// Buffered items from a previous chunk. Items in this list should not have
+  /// been read by the user.
+  List<T> _buffered;
+
+  /// Instance variable representing an empty list object, used as the empty
+  /// default state for [_buffered]. Take caution not to write code that
+  /// directly modify the [_buffered] list by adding elements to it.
+  final List<T> _emptyList = [];
```

Comment:
You migh be able to use a constant list... then writing to it is not possible.
```suggestion
  final List<T> _emptyList = const <T>[];
```

Reply:
This produces an error :O 

```
Constant list literals can't include a type parameter as a type argument, such as 'T'.
Try replacing the type parameter with a different type.dart(invalid_type_argument_in_const_literal)
```

Reply:
ah,